### PR TITLE
Platform Views - Android and MethodChannel API, and implementation skeleton

### PR DESCRIPTION
**Pull request by amirh**
To keep the scope of this CL as small of possible I'm leaving the actual
implementation of the platform view mechanics to a following CL.
This is mostly a noop skeleton.

This CL introduces:
  * A PlatformViewsController class which will be responsible for creating,
    resizing, and disposing platform views.
  * A PlatformViewRegistry which is exposed through the PluginRegistry
    and allows plugins to register factories for platform views.

Android plugin code will add support for a new platform view type by
implementing PlatformViewFactory, and registering a factory object with the
registry, e.g:
```java
registrar.platformViewRegistry().registerViewFactory(
  'webview',
  new FlutterWebViewFactory()
);
```
On the Dart side, the framework will ask the engine to create new
platform views by sending a create message over the platformviews method
channel with the unique platform view type id, dimensions, and a unique
id allocated by the framework for the new platform view instance.
The platformviews method channel is also used for resizing and disposing
platform views.

flutter/flutter#19030

### CODE REVIEW

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    private FlutterView mFlutterView;
+
+    public PlatformViewsController() {
+        mRegistry = new PlatformViewRegistryImpl();
+    }
+
+    public void attachFlutterView(FlutterView view) {
+        if (mFlutterView != null)
+            throw new AssertionError(
+                    "A PlatformsViewController can only be attached to a single FlutterView.\n" +
```
Comment (jason-simmons) [inline shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java]: typo: "PlatformViewsController"

File: travis/licenses_golden/licenses_flutter [hunk 2]
```diff
@@ -202,6 +202,11 @@ FILE: ../../../flutter/lib/ui/window/window.h
 FILE: ../../../flutter/shell/common/skia_event_tracer_impl.cc
 FILE: ../../../flutter/shell/platform/android/apk_asset_provider.cc
 FILE: ../../../flutter/shell/platform/android/io/flutter/plugin/common/JSONUtil.java
+FILE: ../../../flutter/shell/platform/android/io/flutter/plugin/platform/PlatformView.java
```
Comment (jason-simmons) [inline travis/licenses_golden/licenses_flutter]: These paths should be moved to a different position in the license file (check the Travis logs)

**Comment by chinmaygarde** [comment]
Since this feature is not complete yet, is there an internal namespace we can put this in to indicate the lack of stability of these APIs? In the past, APIs created for prototypes have been marked public. On iOS, we just don't publish the headers for the internal Objective-C APIs.

**Comment by amirh** [comment]
@chinmaygarde I'm not sure how to prevent people from using these APIs before the implementation is complete (I mean other than the fact that it won't do anything).
While I can probably make the PlatformViewsRegistry package private, I still need to return something from the PluginRegistry which is completely public...
As for the method channel API, not sure what can be done to hide it...

That being said I personally don't expect people to start relying on this before it is doing something...

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java (hunk 1 above)
Comment (amirh) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java]: done

File: travis/licenses_golden/licenses_flutter (hunk 2 above)
Comment (amirh) [reply travis/licenses_golden/licenses_flutter]: regenerated the licenses files... not sure how did this happen

**Comment by amirh** [comment]
AndroidView is not yet ready to be used.
You can follow the referenced issue for progress.

**Pull request merged**